src/VisualStudio/Core/Def/Packaging/PackageSearchService.cs

```diff
@@ -0,0 +1,614 @@
+            //
+            //      1) If the file is not cached locally it will download it in the background.
+            //         Until the file is downloaded, null will be returned from client.ReadFile.
+            //      2) If the file is cached locally and was downloaded less than (24 * 60) 
+            //         minutes ago, then the client will do nothing (until that time has elapsed).
+            //         Calls to client.ReadFile will return the cached file.
+            //      3) If the file is cached locally and was downloaded more than (24 * 60) 
+            //         minutes ago, then the client will attempt to download the file.
+            //         In the interim period null will be returned from client.ReadFile.
+            var pollingMinutes = (int)TimeSpan.FromDays(1).TotalMinutes;
+            using (var client = _remoteControlService.CreateClient(HostId, serverPath, pollingMinutes))
+            {
+                LogInfo("Creating download client completed");
+
+                // Poll the client every minute until we get the file.
+                while (true)
+                {
+                    _cancellationToken.ThrowIfCancellationRequested();
+
+                    var resultOpt = await TryDownloadFileAsync(client).ConfigureAwait(false);
+                    if (resultOpt == null)
+                    {
+                        var delay = _delayService.CachePollDelay;
+                        LogInfo($"File not downloaded. Trying again in {delay}");
+                        await Task.Delay(delay, _cancellationToken).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        // File was downloaded.  
+                        return resultOpt;
+                    }
+                }
+            }
+        }
+
+        /// <summary>Returns 'null' if download is not available and caller should keep polling.</summary>
+        private async Task<XElement> TryDownloadFileAsync(IPackageSearchRemoteControlClient client)
+        {
+            LogInfo("Read file from client");
+
+            // "ReturnsNull": Only return a file if we have it locally *and* it's not older than our polling time (1 day).
+
+            using (var stream = await client.ReadFileAsync(__VsRemoteControlBehaviorOnStale.ReturnsNull).ConfigureAwait(false))
+            {
+                if (stream == null)
+                {
+                    LogInfo("Read file completed. Client returned no data");
+                    return null;
+                }
+
+                LogInfo("Read file completed. Client returned data");
+                LogInfo("Converting data to XElement");
+                var result = XElement.Load(stream);
+                LogInfo("Converting data to XElement completed");
+                return result;
+            }
+        }
+
+        private async Task RepeatIOAsync(Action action)
+        {
+            const int repeat = 6;
```

Comment:
I'm okay with the repeat count being a perfect number, but I'd prefer it be odd. :wink: